Replace ZeroClipboard with clipboard.js.

XhmikosR opened the pull request:
This is a very first pass at clipboard.js.

It seems to work fine on Firefox.

Things we need to consider:
1. clear the selection after copy
2. add messages for success, error, etc.
3. doesn't seem to work on IE 10
4. doesn't seem to work on Chrome either; throws `Discontiguous selection is not supported.`

/CC @twbs/team 

Fixes #17736.

Closes #17340 #17341.

PS. if anyone can fix any of the todo's please leave a comment after you push so that I don't accidentally force push the branch.

zenorocha commented:
Hey @XhmikosR,

Here's what you need to fix all items listed:

```js
// Insert copy to clipboard button before .highlight
$('.highlight').each(function () {
  var btnHtml = '<div class="bd-clipboard"><span class="btn-clipboard" title="Copy to clipboard">Copy</span></div>'
  $(this).before(btnHtml)
  $('.btn-clipboard').tooltip()
})

var clipboard = new Clipboard('.btn-clipboard', {
  target: function (trigger) {
    return trigger.parentNode.nextElementSibling;
  }
});

clipboard.on('success', function (e) {
  $(e.trigger)
    .tooltip('dispose')
    .attr('title', 'Copied!')
    .tooltip('show')
    .attr('data-original-title', 'Copy to clipboard')

  e.clearSelection();
});

clipboard.on('error', function (e) {
  var fallbackMsg = /Mac/i.test(navigator.userAgent) ? 'Press ⌘-C to copy' : 'Press Ctrl-C copy';

  $(e.trigger)
    .tooltip('dispose')
    .attr('title', fallbackMsg)
    .tooltip('show')
    .attr('data-original-title', 'Copy to clipboard')
});
```

Also, you'll need to update clipboard.js to [v1.4.2](https://github.com/zenorocha/clipboard.js/releases/tag/v1.4.2).

zenorocha commented:
Also, you'll need to update clipboard.js to [v1.4.2](https://github.com/zenorocha/clipboard.js/releases/tag/v1.4.2).

XhmikosR committed:
Replace ZeroClipboard with clipboard.js.

Things left:
1. clear the selection after copy
2. add messages for success and error

zenorocha committed:
Fix the remaining issues for the clipboard.js switch.

XhmikosR committed:
Run `grunt`.

XhmikosR commented:
@zenorocha: many thanks! If you want me to change the commit message of your patch let me know.

@twbs/team: branch updated and works on IE10, FF and Chrome. Please test and confirm.

XhmikosR commented on docs/assets/js/src/application.js:
```diff
@@ -36,50 +36,38 @@
+    });
+
+    clipboard.on('error', function (e) {
+      var fallbackMsg = /Mac/i.test(navigator.userAgent) ? 'Press ⌘-C to copy' : 'Press Ctrl-C copy'
```
Note to self: fix `Press Ctrl-C copy`

cvrebert commented on docs/assets/js/src/application.js:
```diff
@@ -36,50 +36,38 @@
+    });
+
+    clipboard.on('error', function (e) {
+      var fallbackMsg = /Mac/i.test(navigator.userAgent) ? 'Press ⌘-C to copy' : 'Press Ctrl-C to copy'
```
Perhaps we could use `\u2318` instead of a literal `⌘` so as to keep this file in ASCII?

XhmikosR replied:
I agree. I've been meaning to do that but it got late. I'll get to it
tomorrow.
On Oct 14, 2015 01:32, "Chris Rebert" <notifications@github.com> wrote:

> In docs/assets/js/src/application.js
> <https://github.com/twbs/bootstrap/pull/17751#discussion_r41934913>:
>
> > +        return trigger.parentNode.nextElementSibling
> > +      }
> > +    });
> > +
> > +    clipboard.on('success', function (e) {
> > +      $(e.trigger)
> > +        .tooltip('dispose')
> > +        .attr('title', 'Copied!')
> > +        .tooltip('show')
> > +        .attr('data-original-title', 'Copy to clipboard')
> > +
> > +      e.clearSelection()
> > +    });
> > +
> > +    clipboard.on('error', function (e) {
> > +      var fallbackMsg = /Mac/i.test(navigator.userAgent) ? 'Press ⌘-C to copy' : 'Press Ctrl-C to copy'
>
> Perhaps we could use \u2318 instead of a literal ⌘ so as to keep this
> file in ASCII?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twbs/bootstrap/pull/17751/files#r41934913>.
>

zenorocha commented:
LGTM :+1:

XhmikosR commented:
One thing I noticed is that after success, if you hover again over the same copy, the tooltip is the one from success.

zenorocha commented:
Hey @XhmikosR, I tried using `fixTitle` but no success :/

zenorocha commented:
Just noticed that `fixTitle` has been renamed to `_fixTItle`. Here's how it would be using that function.

```js
$('.highlight').each(function () {
  var btnHtml = '<div class="bd-clipboard"><span class="btn-clipboard" title="Copy to clipboard">Copy</span></div>'
  $(this).before(btnHtml)
  $('.btn-clipboard').tooltip()
})

var clipboard = new Clipboard('.btn-clipboard', {
  target: function (trigger) {
    return trigger.parentNode.nextElementSibling
  }
})

clipboard.on('success', function (e) {
  $(e.trigger)
    .attr('title', 'Copied!')
    .tooltip('_fixTitle')
    .tooltip('show')
    .attr('title', 'Copy to clipboard')
    .tooltip('_fixTitle')

  e.clearSelection()
})

clipboard.on('error', function (e) {
  var fallbackMsg = /Mac/i.test(navigator.userAgent) ? 'Press \u2318 to copy' : 'Press Ctrl-C to copy'

  $(e.trigger)
    .attr('title', fallbackMsg)
    .tooltip('_fixTitle')
    .tooltip('show')
    .attr('title', 'Copy to clipboard')
    .tooltip('_fixTitle')
})
```

XhmikosR commented:
Thanks a lot @zenorocha! I think everything is ready to merge now.

XhmikosR commented:
@zenorocha: BTW, you should add a banner in the minified dist file. Now the version, license etc aren't shown.

zenorocha commented:
Yes! working on it https://github.com/zenorocha/clipboard.js/issues/88

XhmikosR commented:
Cool, awesome!

zenorocha commented:
Ok, we put a banner ;)

https://raw.githubusercontent.com/zenorocha/clipboard.js/master/dist/clipboard.min.js

XhmikosR commented:
Cool I'll update the file here too.
On Oct 15, 2015 22:54, "Zeno Rocha" <notifications@github.com> wrote:

> Ok, we put a banner ;)
>
>
> https://raw.githubusercontent.com/zenorocha/clipboard.js/master/dist/clipboard.min.js
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twbs/bootstrap/pull/17751#issuecomment-148502552>.
>

zenorocha commented:
Any news on this?

XhmikosR commented:
I think I'm gonna merge it in the next days as is since I don't seem to get any feedback from the others.

zenorocha commented:
lol

XhmikosR commented:
Something went wrong with the PR?
On Oct 24, 2015 20:45, "Zeno Rocha" <notifications@github.com> wrote:

> lol
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twbs/bootstrap/pull/17751#issuecomment-150836699>.
>

zenorocha commented:
Nope, I'm just happy about it ;)

XhmikosR commented:
Oh definitely. When I found out your project I decided I definitely wanted
it here. Cudos!
On Oct 24, 2015 20:56, "Zeno Rocha" <notifications@github.com> wrote:

> Nope, I'm just happy about it ;)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twbs/bootstrap/pull/17751#issuecomment-150837200>.
>
